redash/query_runner/prometheus.py

```diff
@@ -40,52 +89,66 @@ def get_schema(self, get_stats=False):
         return schema.values()
 
     def run_query(self, query, user):
+        """
+        Query Syntax, actually it is the URL query string.
+        Check the Prometheus HTTP API for the details of the supported query string.
+
+        https://prometheus.io/docs/prometheus/latest/querying/api/
+
+        example: instant query
+            query=http_requests_total
+
+        example: range query
+            query=http_requests_total&start=2018-01-20T00:00:00.000Z&end=2018-01-25T00:00:00.000Z&step=60s
+
+        example: until now range query
+            query=http_requests_total&start=2018-01-20T00:00:00.000Z&step=60s
+            query=http_requests_total&start=2018-01-20T00:00:00.000Z&end=now&step=60s
+        """
+
         base_url = self.configuration["url"]
 
         try:
             error = None
             query = query.strip()
+            # for backward compatibility
+            query = 'query={}'.format(query) if not query.startswith('query=') else query
 
-            local_query = '/api/v1/query'
-            url = base_url + local_query
-            payload = {'query': query}
-            response = requests.get(url, params=payload)
+            payload = parse_qs(query)
+            query_type = 'query_range' if 'step' in payload.keys() else 'query'
+
+            # for the range of until now
+            if query_type == 'query_range' and ('end' not in payload.keys() or 'now' in payload['end']):
+                date_now = datetime.now()
+                payload.update({"end": [date_now.isoformat("T") + "Z"]})
+
+            api_endpoint = base_url + '/api/v1/{}'.format(query_type)
+
+            response = requests.get(api_endpoint, params=payload)
             response.raise_for_status()
-            raw_data = response.json()['data']['result']
-            columns = [
-                {
-                    'friendly_name': 'timestamp',
-                    'type': TYPE_DATETIME,
-                    'name': 'timestamp'
-                },
-                {
-                    'friendly_name': 'value',
-                    'type': TYPE_STRING,
-                    'name': 'value'
-                },
-            ]
-            columns_name = raw_data[0]['metric'].keys()
-            for column_name in columns_name:
-                columns.append({
-                    'friendly_name': column_name,
+
+            metrics = response.json()['data']['result']
+
+            if len(metrics) == 0:
+                return None, 'query result is empty.'
+
+            metric_labels = metrics[0]['metric'].keys()
+
+            for label_name in metric_labels:
+                self.columns.append({
```

Comment:
While usually we don't reuse a query runner, this is not guaranteed behavior, so it's better to have `columns` to be local to the function instead of a class variable.

Reply:
fixed.